pkg/operator/configobserver/cloudconfig/sync_cloud_config.go

```diff
@@ -0,0 +1,121 @@
+)
+
+const (
+	configNamespace                 = "openshift-config"
+	machineSpecifiedConfigNamespace = "openshift-config-managed"
+	machineSpecifiedConfig          = "kube-cloud-config"
+)
+
+var validCloudProviders = sets.NewString("aws", "azure", "gce", "openstack", "vsphere")
+
+// InfrastructureLister lists infrastrucre information and allows resources to be synced
+type InfrastructureLister interface {
+	InfrastructureLister() configlistersv1.InfrastructureLister
+	ResourceSyncer() resourcesynccontroller.ResourceSyncer
+	ConfigMapLister() corelisterv1.ConfigMapLister
+}
+
+// NewCloudConfigSyncer returns a new cloudprovider dependent observer for syncing cloud config per provider
+// for controller-manager and api-server.
+func NewCloudConfigSyncer(targetNamespaceName string) configobserver.ObserveConfigFunc {
+	cloudObserver := &cloudConfigSyncer{
+		targetNamespaceName: targetNamespaceName,
+	}
+	return cloudObserver.SyncCloudConfig
+}
+
+type cloudConfigSyncer struct {
+	targetNamespaceName string
+}
+
+// ObserveCloudProviderNames observes the cloud provider from the global cluster infrastructure resource.
+func (c *cloudConfigSyncer) SyncCloudConfig(genericListers configobserver.Listers, recorder events.Recorder, existingConfig map[string]interface{}) (map[string]interface{}, []error) {
+	listers := genericListers.(InfrastructureLister)
+
+	infrastructure, err := listers.InfrastructureLister().Get("cluster")
+	if err != nil {
+		if errors.IsNotFound(err) {
+			recorder.Warningf("ObserveCloudProviderNames", "Required infrastructures.%s/cluster not found", configv1.GroupName)
+		}
+		return existingConfig, []error{err}
+	}
+
+	cloudProvider := getPlatformName(infrastructure.Status.Platform, recorder)
+
+	// we copy cloudprovider configmap values only for some cloud providers.
+	if !validCloudProviders.Has(cloudProvider) {
+		return existingConfig, nil
+	}
+
+	sourceCloudConfigMap := infrastructure.Spec.CloudConfig.Name
+	sourceCloudConfigNamespace := configNamespace
+
+	// If a managed cloud-provider config is available, it should be used instead of the default. If the configmap is not
+	// found, the default values should be used.
+	if _, err = listers.ConfigMapLister().ConfigMaps(machineSpecifiedConfigNamespace).Get(machineSpecifiedConfig); err == nil {
+		sourceCloudConfigMap = machineSpecifiedConfig
+		sourceCloudConfigNamespace = machineSpecifiedConfigNamespace
+	} else if !errors.IsNotFound(err) {
+		return existingConfig, []error{err}
+	}
+
+	sourceLocation := resourcesynccontroller.ResourceLocation{
+		Namespace: sourceCloudConfigNamespace,
+		Name:      sourceCloudConfigMap,
+		Provider:  cloudProvider,
+	}
+
+	if err := listers.ResourceSyncer().SyncConfigMap(
+		resourcesynccontroller.ResourceLocation{
+			Namespace: c.targetNamespaceName,
+			Name:      "cloud-config",
+			Provider:  cloudProvider,
+		},
+		sourceLocation); err != nil {
+		return existingConfig, []error{err}
+	}
+
+	return existingConfig, nil
```

Comment:
so, it seems like we shouldn't return the existingConfig, but what is the preferred action here?

it seems like this function _wants_ to return existingConfig back to the caller, could we maybe turn this into a true/false return or something so that the caller knows to just use the value it provided?

but, since it seems like this function is mainly updating the configmap, is there a better place for this to live?